inference-engine/src/gna_plugin/memory/gna_memory_state.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "gna_memory_state.hpp"
+#include "frontend/quantized_layer_params.hpp"
+#include "layer_transform.hpp"
+#include "preprocessing.hpp"
+#include "ie_layouts.h"
+
+namespace  GNAPluginNS {
+
+namespace memory {
+
+    std::string GNAMemoryState::GetName() const {
+        return name;
+    }
+
+    void GNAMemoryState::Reset() {
+        state->Reset();
+    }
+
+    InferenceEngine::Precision GNAMemoryState::getPrecision() const {
+        InferenceEngine::Precision state_precision;
+
+        if (state->getInput()) {
+            state_precision = state->getInput()->precision;
+        } else {
+            auto element_size = state->elementSizeBytes();
+            switch (element_size) {
+            case 4:
+                state_precision = InferenceEngine::Precision::FP32;
+                break;
+            case 2:
+                state_precision = InferenceEngine::Precision::I16;
+                break;
+            default:
+                THROW_GNA_EXCEPTION << "Incorrect state element size to determine precision";
+            }
+        }
+
+        return state_precision;
+    }
+
+    void GNAMemoryState::SetState(InferenceEngine::Blob::Ptr newState) {
+        auto data_ptr = newState->cbuffer().as<void*>();
+        auto data_size = newState->byteSize();
+        auto data_elements = data_size / newState->element_size();
+        if (ALIGN64(state->reserved_size) != ALIGN64((data_size / (newState->element_size() / state->elementSizeBytes())))) {
+            THROW_GNA_EXCEPTION << "Failed to SetState. Sizes of new and old states do not match";
+        }
+
+        InferenceEngine::Precision state_precision = getPrecision();
+        auto new_state_precision = newState->getTensorDesc().getPrecision();
+
+        if (new_state_precision == state_precision) {
+            std::memcpy(state->gna_ptr, data_ptr, data_size);
```

Comment:
can we avoid that copy? like getlastState returned a blob ptr, than we modify that blob and call set state - please check that blobPtr are different and copy only in that case